# Data  Wrangling
---
Data Wrangling is a way of transforming raw data into a more readable form.

Examples:
1. Preparing data for input to a machine learning algorithm.
2. Analyzing algorithm output data: transform the data into more understandable formats for graphical uses, etc.

#### Establish a file path reference for the repo and user files:

In [1]:
repo_path = './04_repos-dump.csv'
users_path = './04_user-geocodes-dump.csv'

#### Pandas and Data Set import:

In [2]:
import pandas as pd

# Import Data
users = pd.read_csv(users_path, quotechar = '"', skipinitialspace = True)
repos = pd.read_csv(repo_path, quotechar = '"', skipinitialspace = True)

#### Repo Preview:

In [3]:
display(repos.head())

full_name  stars  forks  \
0  thedaviddias/Front-End-Checklist  24267   2058   
1            GoogleChrome/puppeteer  21976   1259   
2             parcel-bundler/parcel  13981    463   
3   Chalarangelo/30-seconds-of-code  13466   1185   
4      wearehive/project-guidelines  11279    970   

                                         description    language  
0  🗂 The perfect Front-End Checklist for modern w...  JavaScript  
1                           Headless Chrome Node API  JavaScript  
2  📦🚀 Blazing fast, zero configuration web applic...  JavaScript  
3  Curated collection of useful Javascript snippe...  JavaScript  
4    A set of best practices for JavaScript projects  JavaScript

#### Users Preview:

In [4]:
display(users.head())

id             name  type           location        lat  \
0      shprink    Julien Renaux  User   Toulouse, France  43.604652   
1   lllyasviel              NaN  User                NaN        NaN   
2       arzzen     Lukáš Mešťan  User   Zilina, Slovakia  49.219450   
3   javierbyte   Javier Bórquez  User    Guadalajara, MX  20.659699   
4  nbarbettini  Nate Barbettini  User  San Francisco, CA  37.774929   

         long           city        country  
0    1.444209       Toulouse         France  
1         NaN            NaN            NaN  
2   18.740800         Žilina       Slovakia  
3 -103.349609    Guadalajara         Mexico  
4 -122.419415  San Francisco  United States

#### Drop duplicate entries, split "full_name" into 2 columns, and rename "id" column to "users"
 - Drop duplicates from the repos list
 - Since the shape before and shape after are the same, there were no duplicates.

In [5]:
print(f"Shape before dropping duplicates: {repos.shape}\n")
repos = repos.drop_duplicates(subset='full_name', keep='last')

print(f"Shape after dropping duplicates: {repos.shape}\n")
display(repos.head())

Shape before dropping duplicates: (8697, 5)

Shape after dropping duplicates: (8697, 5)



full_name  stars  forks  \
0  thedaviddias/Front-End-Checklist  24267   2058   
1            GoogleChrome/puppeteer  21976   1259   
2             parcel-bundler/parcel  13981    463   
3   Chalarangelo/30-seconds-of-code  13466   1185   
4      wearehive/project-guidelines  11279    970   

                                         description    language  
0  🗂 The perfect Front-End Checklist for modern w...  JavaScript  
1                           Headless Chrome Node API  JavaScript  
2  📦🚀 Blazing fast, zero configuration web applic...  JavaScript  
3  Curated collection of useful Javascript snippe...  JavaScript  
4    A set of best practices for JavaScript projects  JavaScript

#### Divide the full_name into 2 separate columns:
 - in the repos list, the full_name format is currently: username/repo_name
 - Desired Output: 1st column is the user name, 2nd column is the repo name

In [6]:
def extract_user(line):
  return line.split('/')[0]

# Alternate:
#def extract_user(line):
#  line = line.split('/')
#  return line[0]

def extract_repo(line):
  return line.split('/')[1]

# This will parse and make 2 new columns:
repos['user'] = repos['full_name'].str[:].apply(extract_user)
repos['repo'] = repos['full_name'].str[:].apply(extract_repo)

print(f"Shape After Processing: {repos.shape}\n")
display(repos.head())

# Rows are the same as before, but 2 additional columns have been added

Shape After Processing: (8697, 7)



full_name  stars  forks  \
0  thedaviddias/Front-End-Checklist  24267   2058   
1            GoogleChrome/puppeteer  21976   1259   
2             parcel-bundler/parcel  13981    463   
3   Chalarangelo/30-seconds-of-code  13466   1185   
4      wearehive/project-guidelines  11279    970   

                                         description    language  \
0  🗂 The perfect Front-End Checklist for modern w...  JavaScript   
1                           Headless Chrome Node API  JavaScript   
2  📦🚀 Blazing fast, zero configuration web applic...  JavaScript   
3  Curated collection of useful Javascript snippe...  JavaScript   
4    A set of best practices for JavaScript projects  JavaScript   

             user                 repo  
0    thedaviddias  Front-End-Checklist  
1    GoogleChrome            puppeteer  
2  parcel-bundler               parcel  
3    Chalarangelo   30-seconds-of-code  
4       wearehive   project-guidelines

#### Rename the `id` column to `user`

In [7]:
display(users.head())

users.rename(columns = {'id': 'user'}, inplace = True)  # renames column

display(users.head())

id             name  type           location        lat  \
0      shprink    Julien Renaux  User   Toulouse, France  43.604652   
1   lllyasviel              NaN  User                NaN        NaN   
2       arzzen     Lukáš Mešťan  User   Zilina, Slovakia  49.219450   
3   javierbyte   Javier Bórquez  User    Guadalajara, MX  20.659699   
4  nbarbettini  Nate Barbettini  User  San Francisco, CA  37.774929   

         long           city        country  
0    1.444209       Toulouse         France  
1         NaN            NaN            NaN  
2   18.740800         Žilina       Slovakia  
3 -103.349609    Guadalajara         Mexico  
4 -122.419415  San Francisco  United States

user             name  type           location        lat  \
0      shprink    Julien Renaux  User   Toulouse, France  43.604652   
1   lllyasviel              NaN  User                NaN        NaN   
2       arzzen     Lukáš Mešťan  User   Zilina, Slovakia  49.219450   
3   javierbyte   Javier Bórquez  User    Guadalajara, MX  20.659699   
4  nbarbettini  Nate Barbettini  User  San Francisco, CA  37.774929   

         long           city        country  
0    1.444209       Toulouse         France  
1         NaN            NaN            NaN  
2   18.740800         Žilina       Slovakia  
3 -103.349609    Guadalajara         Mexico  
4 -122.419415  San Francisco  United States

#### Merging The Datasets:
 - use `merge` function in pandas to merge the 2 datasets
 - merge `repos` and `users` according to the name of the user, according to the left value (repos).
 - There are 8697 repos and 6246 users.
 - Notice that the # of columns now is 14 since the datasets are merged:

In [8]:
repos_users = pd.merge(repos, users, on = 'user', how = 'left')

# 14 columns
print(f'Shape After Merge: {repos_users.shape}\n')
display(repos_users.head())

Shape After Merge: (8697, 14)



full_name  stars  forks  \
0  thedaviddias/Front-End-Checklist  24267   2058   
1            GoogleChrome/puppeteer  21976   1259   
2             parcel-bundler/parcel  13981    463   
3   Chalarangelo/30-seconds-of-code  13466   1185   
4      wearehive/project-guidelines  11279    970   

                                         description    language  \
0  🗂 The perfect Front-End Checklist for modern w...  JavaScript   
1                           Headless Chrome Node API  JavaScript   
2  📦🚀 Blazing fast, zero configuration web applic...  JavaScript   
3  Curated collection of useful Javascript snippe...  JavaScript   
4    A set of best practices for JavaScript projects  JavaScript   

             user                 repo              name          type  \
0    thedaviddias  Front-End-Checklist        David Dias          User   
1    GoogleChrome            puppeteer               NaN  Organization   
2  parcel-bundler               parcel            Parcel  Organization   
3    Chalarangelo   30-seconds-of-code  Angelos Chalaris          User   
4       wearehive   project-guidelines              Hive  Organization   

                    location        lat       long    city         country  
0  France, Mauritius, Canada        NaN        NaN     NaN             NaN  
1                        NaN        NaN        NaN     NaN             NaN  
2                        NaN        NaN        NaN     NaN             NaN  
3             Athens, Greece  37.983810  23.727539  Athens          Greece  
4                     London  51.507351  -0.127758  London  United Kingdom

#### Rearranging the columns of merged list: Reindexing

In [9]:
repos_users = repos_users.reindex(['user', 'repo', 'description', 'stars', 'forks', 'language', 'full_name', 'name', 'type', 'location', 'lat', 'long', 'city', 'country'], axis = 1)
# 14 columns

print(f'Shape after Reindexing: {repos_users.shape}\n')

display(repos_users.head())

Shape after Reindexing: (8697, 14)



user                 repo  \
0    thedaviddias  Front-End-Checklist   
1    GoogleChrome            puppeteer   
2  parcel-bundler               parcel   
3    Chalarangelo   30-seconds-of-code   
4       wearehive   project-guidelines   

                                         description  stars  forks  \
0  🗂 The perfect Front-End Checklist for modern w...  24267   2058   
1                           Headless Chrome Node API  21976   1259   
2  📦🚀 Blazing fast, zero configuration web applic...  13981    463   
3  Curated collection of useful Javascript snippe...  13466   1185   
4    A set of best practices for JavaScript projects  11279    970   

     language                         full_name              name  \
0  JavaScript  thedaviddias/Front-End-Checklist        David Dias   
1  JavaScript            GoogleChrome/puppeteer               NaN   
2  JavaScript             parcel-bundler/parcel            Parcel   
3  JavaScript   Chalarangelo/30-seconds-of-code  Angelos Chalaris   
4  JavaScript      wearehive/project-guidelines              Hive   

           type                   location        lat       long    city  \
0          User  France, Mauritius, Canada        NaN        NaN     NaN   
1  Organization                        NaN        NaN        NaN     NaN   
2  Organization                        NaN        NaN        NaN     NaN   
3          User             Athens, Greece  37.983810  23.727539  Athens   
4  Organization                     London  51.507351  -0.127758  London   

          country  
0             NaN  
1             NaN  
2             NaN  
3          Greece  
4  United Kingdom

#### Ranking based on the "stars" column
 - adds a new column to the end without modifying the data

In [10]:
repos_users['rank'] = repos_users['stars'].rank(ascending = False)

print(repos_users.shape)  # 14 columns

display(repos_users.head())

(8697, 15)


user                 repo  \
0    thedaviddias  Front-End-Checklist   
1    GoogleChrome            puppeteer   
2  parcel-bundler               parcel   
3    Chalarangelo   30-seconds-of-code   
4       wearehive   project-guidelines   

                                         description  stars  forks  \
0  🗂 The perfect Front-End Checklist for modern w...  24267   2058   
1                           Headless Chrome Node API  21976   1259   
2  📦🚀 Blazing fast, zero configuration web applic...  13981    463   
3  Curated collection of useful Javascript snippe...  13466   1185   
4    A set of best practices for JavaScript projects  11279    970   

     language                         full_name              name  \
0  JavaScript  thedaviddias/Front-End-Checklist        David Dias   
1  JavaScript            GoogleChrome/puppeteer               NaN   
2  JavaScript             parcel-bundler/parcel            Parcel   
3  JavaScript   Chalarangelo/30-seconds-of-code  Angelos Chalaris   
4  JavaScript      wearehive/project-guidelines              Hive   

           type                   location        lat       long    city  \
0          User  France, Mauritius, Canada        NaN        NaN     NaN   
1  Organization                        NaN        NaN        NaN     NaN   
2  Organization                        NaN        NaN        NaN     NaN   
3          User             Athens, Greece  37.983810  23.727539  Athens   
4  Organization                     London  51.507351  -0.127758  London   

          country  rank  
0             NaN   3.0  
1             NaN   4.0  
2             NaN  11.0  
3          Greece  13.0  
4  United Kingdom  16.0

#### How many Python users are in the new 'repos_users' file?
 - Display the shape of just the python users.
 - Display the dataframe with only python users.

In [11]:
print(repos_users[repos_users['language'] == 'Python'].shape)

# display only the Python users:
repos_users[repos_users['language'] == 'Python'].head()

(1357, 15)


user                  repo  \
3308      donnemartin  system-design-primer   
3309           python               cpython   
3310         ageitgey      face_recognition   
3311  tonybeltramelli              pix2code   
3312           google           python-fire   

                                            description  stars  forks  \
3308  Learn how to design large-scale systems. Prep ...  21780   2633   
3309                    The Python programming language  15060   3779   
3310  The world's simplest facial recognition api fo...   8487   1691   
3311  pix2code: Generating Code from a Graphical Use...   8037    605   
3312  Python Fire is a library for automatically gen...   7663    386   

     language                         full_name              name  \
3308   Python  donnemartin/system-design-primer      Donne Martin   
3309   Python                    python/cpython            Python   
3310   Python         ageitgey/face_recognition      Adam Geitgey   
3311   Python          tonybeltramelli/pix2code  Tony Beltramelli   
3312   Python                google/python-fire            Google   

              type          location        lat       long        city  \
3308          User  Washington, D.C.  38.907192 -77.036871  Washington   
3309  Organization               NaN        NaN        NaN         NaN   
3310          User    Various places        NaN        NaN         NaN   
3311          User           Denmark  56.263920   9.501785         NaN   
3312  Organization               NaN        NaN        NaN         NaN   

            country  rank  
3308  United States   5.0  
3309            NaN   9.0  
3310            NaN  31.0  
3311        Denmark  34.0  
3312            NaN  36.0